lib/src/web_socket/common/w_socket.dart

```diff
@@ -27,20 +32,71 @@ abstract class CommonWSocket extends Stream implements WSocket {
   /// no close reason available this property will be `null`.
   String closeReason;
 
+  /// The error that was either added to the socket sink or received from the
+  /// socket stream. This will only be populated if an error occurs.
+  var error;
+
+  /// Incoming communication. Data from the web socket will be piped
+  /// to this controller's sink. The stream will be exposed, allowing
+  /// users to listen to the incoming data.
+  StreamController incoming;
+
   /// Whether or not the web socket is in the process of closing (or already has
   /// closed). This prevents duplicate behavior if [close] is called multiple
   /// times.
-  bool _closed = false;
+  bool isClosed = false;
+
+  /// Whether or not the outgoing sink has been closed.
+  bool isSinkClosed = false;
+
+  /// Whether or not the incoming socket stream connection has been closed.
+  bool isSocketClosed = false;
+
+  /// The stack trace from the error that was either added to the socket sink or
+  /// received from the socket stream. This will only be populated if an error
+  /// occurs.
+  StackTrace stackTrace;
+
+  /// Outgoing communication. Sink will be exposed, allowing users to
+  /// add items to the outgoing stream.
+  StreamController outgoing;
+
+  /// The close event with close code and reason from the socket connection.
+  /// Will only be populated once the socket has actually closed.
+  WSocketCloseEvent _closeEvent;
+
+  /// Completer that will complete when [_allClosed] has completed and the close
+  /// code and reason have been set. If the socket closed due to an error, this
+  /// will complete with an error. Otherwise, it will complete normally.
+  Completer<Null> _done = new Completer();
 
   CommonWSocket() {
-    done.whenComplete(() {
-      _closed = true;
+    // Wait for both the sink and the socket to finish closing before exposing
+    // the close code, close reason, and completing the [_done] completer.
+    // If the socket closed due to an error, the [_done] completer will be
+    // completed with that error.
+    allClosed.future.then((_) {
```

Comment:
I had then public because I thought the extending classes need them, but I ended up moving more code than I originally thought into the common class, so only `incoming` and `outgoing` need to be public for that reason. The rest I've switched back to private